File: evaluateReversePolishNotation/evaluateReversePolishNotation.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution{
public:
  int evalRPN(vector<string>& tokens) {
    if (tokens.size() == 0)
      return 0;
    stack<int> s;
    for (string token : tokens) {
      if (token == "+") {
	int num1 = s.top();
	s.pop();
	int num2 = s.top();
	s.pop();
	s.push(num1 + num2);
      } else if (token == "-") {
	int num1 = s.top();
	s.pop();
	int num2 = s.top();
	s.pop();
	s.push(num2 - num1);
      } else if (token == "*") {
	int num1 = s.top();
	s.pop();
	int num2 = s.top();
	s.pop();
	s.push(num2 * num1);
      } else if (token == "/") {
	int num1 = s.top();
	s.pop();
	int num2 = s.top();
	s.pop();
	s.push(num2 / num1);
      } else {
	s.push(stoi(token));
      }
    }
    return s.top();
  }
};

int main()
{
  Solution s;
  vector<string> vec = {"4", "13", "5", "/", "+"};
  cout << s.evalRPN(vec);
  cout << endl;
  return 0;
}

File: 150.evaluateReversePolishNotation/evaluateReversePolishNotation.java
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> numStack = new Stack<>();
        for (String token : tokens) {
            if (token.equals("+")) {
                int right = numStack.pop();
                int left = numStack.pop();
                numStack.push(left + right);
            } else if (token.equals("-")) {
                int right = numStack.pop();
                int left = numStack.pop();
                numStack.push(left - right);
            } else if (token.equals("*")) {
                int right = numStack.pop();
                int left = numStack.pop();
                numStack.push(left * right);
            } else if (token.equals("/")) {
                int right = numStack.pop();
                int left = numStack.pop();
                numStack.push(left / right);
            } else {
                numStack.push(Integer.valueOf(token));
            }
        }
        return numStack.peek();
    }
}